lib/Sema/TypeCheckAttr.cpp

```diff
@@ -5513,3 +5514,51 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
     checker.visit(attr);
   }
 }
+
+void AttributeChecker::visitHasAsyncAlternativeAttr(
+    HasAsyncAlternativeAttr *attr) {
+  // TODO: We should lookup and store the referenced function decl now that they
+  //       have all been parsed. From there, we can verify that the two function
+  //       types match.
+
+  // We have two function decls to worry about here for now.
+  // There will be a third for the found async function decl.
+  // We have the decl that the attribute is attached to
+  //  callingFunctionDecl
+  // We have the completion handler decl passed into the function
+  //  completionHandlerDecl
+  auto callingFunctionDecl = cast<AbstractFunctionDecl>(D);
+  if (callingFunctionDecl->hasAsync()) {
+    diagnose(callingFunctionDecl->getAsyncLoc(), diag::pound_error,
+             "'@hasAsyncAlternative' attached to async function");
+    return;
+  }
+
+  const ParameterList *params = callingFunctionDecl->getParameters();
+  if (!params) // TODO: Figure out how to exercise this and emit an error
+    llvm_unreachable("The parameter list is a nullptr");
+
+  // Ensure that this function has enough arguments to actually have a
+  // completion handler in the specified place, and that the of that completion
+  // handler is a function type.
+
+  if (params->size() <= attr->CompletionHandlerIndex) {
+    diagnose(attr->CompletionHandlerIndexLoc, diag::pound_error,
+             "'@hasAsyncAlternative' parameter 'completionHandlerIndex' is out "
+             "of range");
+    return;
+  }
+
+  auto completionHandlerDecl = params->get(attr->CompletionHandlerIndex);
+  AnyFunctionType *handlerType =
+      completionHandlerDecl->getType()->getAs<AnyFunctionType>();
+
+  if (!handlerType) {
+    diagnose(completionHandlerDecl->getLoc(), diag::pound_error,
+             "'@hasAsyncAlternative' parameter 'completionHandlerIndex' "
+             "must be a function type");
+    return;
+  }
+
```

Comment:
Should verify that the return type of the completion handler is a `Void` type.

Reply:
There's some other checks that might be worth doing as well - ie. the handler shouldn't be attributed with `@autoclosure`. We could probably also check that it has `@escaping`?

Reply:
Done